### ISSUE TITLE
More Operator Overloading

### ISSUE THREAD

**Issue by eeeey** [wurstscript/WurstScript#297]
Hi there,

I would like to suggest new language features, namely more operator overloading. It would be nice to be able to overload [] and () operators to create array-like access to data structures and functors. Examples:

``` Wurst
package First
    class MyList<T>
        // Some stuff
        function operator[](int index) returns T
            return getEntry(index).elem

    class MyFunctor
        private real memory = 0.0

        function operator()(real r1, real r2) returns real
            memory += r1*r2
            return memory // Just an example

    init
        // Usage operator[]
        MyList<int> list = new MyList<int>()
        print(list[0].toString())

        // Usage operator()
        MyFunctor func = new MyFunctor()
        print(func(2.0, 3.0).toString()) // Displays 6
        print(func(2.0, 3.0).toString()) // Displays 12
        print(func(2.0, 3.0).toString()) // Displays 18 
endpackage
```

And keep the good work :)

**Comment by blakehawkins** [comment]
I'm not a fan of overloading function declaration, but I do like [].

Can we have overloading for <, >, <=, >= operators?

There is a fairly well-defined convention for comparing tuples, for example:

    public function vec2.op_lt(vec2 other) returns boolean
        return (this.x < other.x) or (this.x == other.x and this.y < other.y)

**Comment by Cokemonkey11** [comment]
I'm not a fan of overloading function declaration, but I do like [].

Can we have overloading for <, >, <=, >= operators?

There is a fairly well-defined convention for comparing tuples, for example:

    public function vec2.op_lt(vec2 other) returns boolean
        return (this.x < other.x) or (this.x == other.x and this.y < other.y)

**Comment by peterzeller** [comment]
I think for the comparison operator it would be nicer to link the operators to a comparable interface or type-class. It would feel strange to overload <, >, <=, >= and == with 5 separate functions.

Defining < for vec2 does not make sense in my opinion, because there is no default linear order. Some people might expect that the comparison is done by length instead of lexicographically.

**Comment by peq** [comment]
I think for the comparison operator it would be nicer to link the operators to a comparable interface or type-class. It would feel strange to overload <, >, <=, >= and == with 5 separate functions.

Defining < for vec2 does not make sense in my opinion, because there is no default linear order. Some people might expect that the comparison is done by length instead of lexicographically.

**Comment by Cokemonkey11** [comment]
> I think for the comparison operator it would be nicer to link the operators to a comparable interface or type-class. It would feel strange to overload <, >, <=, >= and == with 5 separate functions.

Yes, I would be happy with that as well. Seems like more work for you though :P 

> Defining < for vec2 does not make sense in my opinion, because there is no default linear order. Some people might expect that the comparison is done by length instead of lexicographically.

vec2 always has length of 2. For a template type like pair<T, T>, I agree with you, but for vectors of numbers I think the convention I mentioned above is really quite standard.

Python:

    >>> (1, 2) < (1, 3)
    True
    >>> (3, 1) < (2, 5000)
    False
    >>> (1, 2) < (2, 1)
    True

I did think of one exception - partial ordering https://en.wikipedia.org/wiki/Vector_clock#Partial_ordering_property

I think it makes sense for vec2 to have at least *some* default ordering.

**Comment by peq** [comment]
A Python tuple is something more generic. You can use tuples for many different things, so there will be some cases where lexicographical ordering makes senses.

But vec2 is really supposed to be a vector with 2 dimensions and there is no standard ordering for this domain. I also cannot think of an example where you want to ask a question like: "Is the position of unit A less than the position of unit B?". Comparing positions like this just does not make sense.

**Comment by Cokemonkey11** [comment]
Okay, I think that's fair. I argue that vec2 isn't necessarily for representing location of unit, but I can see that for the typical use case, having an ordering could do more harm than good.

**Comment by Frotty** [comment]
index operator can now be overloaded
